SVG plugin additions

elblake opened the pull request:
Additions and fixes to the SVG path plugin.

Changes include:
 * New path tokenizer
 * Handling of more path commands (H V Q S T A h v q s t a)
 * XML parsing with xmerl
 * More SVG elements handled (less need for the user to convert objects to path before saving)
 * Scaling calculated from document size and viewbox to match the scaling of EPS/PS and AI importers.
 * Fixed relative positioning of sub-paths
 * Added clipped mask textures
 * Added path fill color to mesh

dgud commented on plugins_src/import_export/wpc_svg_path.erl:
```diff
@@ -41,34 +65,259 @@
+        true  -> SetViewBoxScale_1 = SetViewBoxScale;
+        false -> SetViewBoxScale_1 = false
+    end,
+    case catch try_import_svg(Name, Nsubsteps, AutoScale, SetScale,
```
Would prefer a `try catch E:R:ST ->` construct here, we try to avoid the old "catch" construct.

dgud commented on plugins_src/import_export/wpc_svg_path.erl:
```diff
@@ -77,82 +326,360 @@ make_svg(Name, Nsubsteps) ->
+                RescaleW = VWSc / DocUnitRescale,
+                RescaleH = VHSc / DocUnitRescale
+        end,
+        io:format("Rescale: ~p,~p ~n~n", [RescaleW,RescaleH]),
```
Remove debug formats

dgud commented on plugins_src/import_export/wpc_svg_path.erl:
```diff
@@ -225,4 +758,2941 @@ getcedges([#pathop{opkind=pcurveto,x1=X1,y1=Y1,x2=X2,y2=Y2,x3=X,y3=Y}|Ops],
+            {ok, CommandList_3} = assign_transforms(CommandList_2),
+            {ok, CommandList_3, #document_s{fullfilename=FullFilename,units=DocUnits,docsz=DocSz,viewbox=ViewBox}};
+        {Error, {_,_,Line}, Reason, _ET, _St} ->
+            io:format("SVG:~p: ERROR: ~p:~p~n", [Line, Error, Reason]),
```
This debug printout is ok, i.e. for errors cases debug printouts are ok but for the normal case it should
be quiet.

dgud commented on plugins_src/import_export/wpc_svg_path.erl:
```diff
@@ -225,4 +758,2941 @@ getcedges([#pathop{opkind=pcurveto,x1=X1,y1=Y1,x2=X2,y2=Y2,x3=X,y3=Y}|Ops],
+substitute_commands([[NumDigit|_]=X,Y | Rest], O,
+    #subst_curv_pos{curr_cmd=lineto_abs}=CurrPos)
+    when ?IS_DIGIT(NumDigit)
+->
```
I don't like to start the line with '->'  
'when' should be indented with 2 spaces to indicate that it belongs to the function header.

If you want I can indent the file before merging.

dgud reviewed (commented):
This also looks really good too me from a quick glance, 
besides some really minor nitpicks.
I have not done any testing (I assume you have).

Do you me to merge it now or are there more commits coming?

elblake commented:
Hello dgud,
I've fixed the indentations of the when keywords and removed the debug formats.
There are no more commits coming, everything is ready for merging.

dgud commented on plugins_src/import_export/wpc_svg_path.erl:
```diff
@@ -319,17 +318,17 @@ make_svg(Name, #svg_importer_params{
         wpa:pref_set(wpc_svg_path, svg_use_viewbox_coords, UseViewboxCoords),
         wpa:pref_set(wpc_svg_path, svg_viewbox_scale, SetViewBoxScale),
         {ok, E3dFile};
-    {error,Reason} ->
-        {error, ?__(1,"Inkscape .svg path import failed")++": " ++ Reason};
     E ->
         io:format("File Import Error Report:\n ~p\n",[E]),
         {error, ?__(2,"Inkscape .svg path import internal error")}
+    catch
+        error:Reason ->
+            {error, ?__(1,"Inkscape .svg path import failed")++": " ++ Reason}
```
It's the **E** branch you will need to catch, the {error, Reason} is a normal return value.

Suggestion:   (ST = StackTrace so we can see where there error occured) 

```
catch Error:Reason:ST ->   
        io:format("File Import Error Report:\n ~p:~p\n  ~p\n",[Error, Reason, ST]),
        {error, ?__(2,"Inkscape .svg path import internal error")}
```

elblake commented:
Oh oops, I've made the change.

dgud commented:
I got the following error when running dialyzer, which I must confess I didn't really understand.

```
1> dialyze().
wpc_svg_path.erl:2330:9: The pattern 
          {image,
           _ImageInfo = {image_tag_r, none, _, _, _, _, _, _, _},
           _ImageSubCmds} can never match the type 
          any()

```

I fixed it by 
modified   plugins_src/import_export/wpc_svg_path.erl
@@ -85,7 +85,7 @@
 
 -record(image_tag_r,
     {
-    clip_path :: string(),
+    clip_path :: string() | none,
     x :: float(),
     y :: float(),
     width :: float(),

Is that ok?

elblake commented:
Yup that is fine, I missed that when I added the type.
